src/connector/src/sink/iceberg.rs

```diff
@@ -375,46 +377,118 @@ impl Sink for IcebergSink {
     }
 }
 
-pub struct IcebergWriter(IcebergWriterEnum);
+pub struct IcebergWriter {
+    inner_writer: IcebergWriterEnum,
+    schema: SchemaRef,
+}
 
 enum IcebergWriterEnum {
-    AppendOnly(AppendOnlyWriter),
-    Upsert(UpsertWriter),
+    AppendOnly(RecordBatchWriter),
+    Upsert(DeltaWriter),
 }
 
 impl IcebergWriter {
-    fn metrics_layer(write_param: &SinkWriterParam) -> WriterPrometheusLayer {
-        let iceberg_metrics = FileAppenderMetrics::new(
-            write_param
+    pub async fn new_append_only(table: Table, writer_param: &SinkWriterParam) -> Result<Self> {
+        let builder_helper = table.builder_helper()?;
+
+        let data_file_builder = DataFileWriterBuilder::new(BaseFileWriterWithMetricsBuilder::new(
+            builder_helper
+                .rolling_writer_builder(builder_helper.parquet_writer_builder(0, None)?)?,
+            writer_param
                 .sink_metrics
-                .iceberg_file_appender_write_qps
+                .iceberg_rolling_unfushed_data_file
                 .deref()
                 .clone(),
-            write_param
+        ));
+        let partition_data_file_builder = FanoutPartitionedWriterWithMetricsBuilder::new(
+            builder_helper.fanout_partition_writer_builder(data_file_builder.clone())?,
+            writer_param
                 .sink_metrics
-                .iceberg_file_appender_write_latency
+                .iceberg_partition_num
                 .deref()
                 .clone(),
         );
-        WriterPrometheusLayer::new(iceberg_metrics)
-    }
-
-    pub async fn new_append_only(table: Table, writer_param: &SinkWriterParam) -> Result<Self> {
-        let metrics_layer = Self::metrics_layer(writer_param);
-        Ok(Self(IcebergWriterEnum::AppendOnly(
-            AppendOnlyWriter::new(table, metrics_layer).await?,
-        )))
+        let dispatch_builder = builder_helper
+            .dispatcher_writer_builder(partition_data_file_builder, data_file_builder)?;
+        // wrap a layer with collect write metrics
+        let prometheus_builder = PrometheusWriterBuilder::new(
+            dispatch_builder,
+            WriterMetrics::new(
+                writer_param.sink_metrics.iceberg_write_qps.deref().clone(),
+                writer_param
+                    .sink_metrics
+                    .iceberg_write_latency
+                    .deref()
+                    .clone(),
```

Comment:
By `deref` and `clone` here, the ref count for label cannot work correctly.

This appears to be unsound. cc @wenym1 Is there a way to prevent this kind of leaking? Or maybe it's another motivation for us to migrate the prometheus library. https://github.com/risingwavelabs/risingwave/issues/13086

Reply:
The `PrometheusWriterBuilder` is defined in icelake, which means we are not able to choose the metric type we expect. If it's using the raw type of metric, there is no way to deal with the issue, even if we migrate to another prometheus library. The only way seems to be forking the current library to embed the current label guard logic in the internal implementation of current library and add a crate patch to the cargo.toml.

@ZENOTME @liurenjie1024 It is possible to let PrometheusWriterBuilder define some metric trait and pass a box dyn trait to it so that it can support customized metric report logic?

Reply:
Or we should the promtheus writer builder in risinwave?

Reply:
> It is possible to let PrometheusWriterBuilder define some metric trait and pass a box dyn trait to it so that it can support customized metric report logic?

Yes, it's possible. 

> By deref and clone here, the ref count for label cannot work correctly.

Actually the semantic here is more like to extract the ref of metrics. The ref count don't need to inc if we can express the ref semantic which means the life time of the ref metrics is subrange to the metrics. So another solution maybe :
- label metrics have an interface to expose the ref of raw metrics.
- PrometheusWriterBuilder can take a ref of raw metrics. (I'm not sure whether it can

Reply:
> This appears to be unsound. cc @wenym1 Is there a way to prevent this kind of leaking? 

It seems hard to prevent because Deref makes it convenient to use the inner type. Except we expose the inner type op (add, set) manually.
```
impl<T, const N: usize> Deref for LabelGuardedMetric<T, N> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
```